### PR TITLE
make ted train on chunks

### PR DESCRIPTION
**Pull request by Ghostvv**
**Proposed changes**:
- closes https://github.com/RasaHQ/rasa/issues/8315
- closes https://github.com/RasaHQ/rasa/issues/8316 

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by koernerfelicia** [comment]
@Ghostvv heads up I'm out starting tomorrow until next Friday

File: rasa/core/featurizers/tracker_featurizers.py [hunk 1]
```diff
@@ -240,6 +247,299 @@ def featurize_trackers(
...
+            for turn_state_features, turn_label_ids, turn_entity_tags in zip(
+                tracker_state_features_chunk, label_ids_chunk, entity_tags_chunk
+            ):
+                example = tf.train.Example(
+                    features=tf.train.Features(
+                        feature=self._to_tf_features(
+                            turn_state_features, turn_label_ids, turn_entity_tags
+                        )
+                    )
+                )
```
Comment (dakshvar22) [inline rasa/core/featurizers/tracker_featurizers.py:445]: Hmm, we should check if we can merge this function and `persist_chunk` inside rasa.shared.nlu.training_data.training_data.py` and create a generic TFRecord writer function that can be reused.

File: rasa/core/featurizers/tracker_featurizers.py [hunk 2]
```diff
@@ -240,6 +247,299 @@ def featurize_trackers(
...
+                # Append each example into tfrecord
+                tfwriter.write(example.SerializeToString())
+
+        return file_path
+
+    def load_chunk(
```
Comment (dakshvar22) [inline rasa/core/featurizers/tracker_featurizers.py:451]: Could you add some comments for future purposes in this function?

File: rasa/model_training.py [hunk 3]
```diff
@@ -590,7 +590,9 @@ async def _train_core_with_validated_data(
                 await rasa.core.train.train_in_chunks(
                     domain_file=domain,
                     training_resource=file_importer,
-                    output_path=Path(_train_path, DEFAULT_CORE_SUBDIRECTORY_NAME),
+                    output_path=os.path.join(
```
Comment (dakshvar22) [inline rasa/model_training.py:593]: Can you leave it to `Path` to be consistent with the code? Any particular reason why it should change?

File: rasa/shared/utils/chunk_utils.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
+import logging
```
Comment (dakshvar22) [inline rasa/shared/utils/chunk_utils.py:1]: We can add the generic persist and load methods for chunk here.

File: rasa/model_training.py [hunk 5]
```diff
@@ -590,7 +590,9 @@ async def _train_core_with_validated_data(
                 await rasa.core.train.train_in_chunks(
```
Comment (dakshvar22) [inline rasa/model_training.py:590]: Do we want to support finetuning?
It's okay to leave it out in case this is going to be a prototype, but i think you will just need to add a `model_to_finetune` parameter to this method and set the agent's ensemble to `model_to_finetune`'s policy ensemble and the rest should be fine.

**Review by dakshvar22 — COMMENTED** [review]
The logic seems largely fine to me at first glance.
I take that the code quality isn't expected to be merge ready and just a prototype?

File: rasa/model_training.py (hunk 3 above)
Comment (Ghostvv) [reply rasa/model_training.py:593]: this is the change from `main`

**Comment by Ghostvv** [comment]
> I take that the code quality isn't expected to be merge ready and just a prototype?

yes

File: rasa/model_training.py (hunk 5 above)
Comment (Ghostvv) [reply rasa/model_training.py:590]: my thinking was: if you do fine-tuning, you don't fine-tune on a lot of data, so normal train should be good

**Comment by koernerfelicia** [comment]
NOTE: there is a known bug (maybe two) with this implementation (see discussion [here](https://rasa-hq.slack.com/archives/C8LEHMTAM/p1619533190265700?thread_ts=1618303780.462500&cid=C8LEHMTAM)).

The first, known bug occurs with datasets that have a long tail distribution of action labels. For example, end-to-end datasets with lots of unique bot utterances (as the label for these is `action_text`).

This means that the chunks are not (even roughly) equal in size. The first chunk will contain examples for each label, and subsequent chunks will be smaller, as there are no more examples for this label. This occurs [here](https://github.com/RasaHQ/rasa/blob/97c8d8d28552cdbab452e4ff61caac216f32a213/rasa/core/featurizers/tracker_featurizers.py#L351-L357).

The second, potential bug, which is possibly related is that the subsequent smaller chunks may not have entities. We hypothesized that this may be because (in this scenario with the long tail of unique bot utterances) the large chunk contains all the unique utterances and the subsequent chunks contain examples only for `action_listen`. It's not clear if this is related, since Vova did provision for missing entities [here](https://github.com/RasaHQ/rasa/blob/97c8d8d28552cdbab452e4ff61caac216f32a213/rasa/core/featurizers/tracker_featurizers.py#L359-L363).

The bug can be reproduced with MultiWOZ for 400 dialogues (312 train, these dialogues are produced with the script [here](https://github.com/RasaHQ/training-data/blob/main/public/MultiWOZ/scripts/dataset_to_rasa.py) with number of dialogues flag set to 400) and chunk=5.
